lib/parsers/acorn-extensions.js

```diff
@@ -15,154 +15,180 @@ exports.enableTolerance = function (parser) {
+  ];
+
+  if (parser.type === tt.comma &&
+      peekNextType(parser) === tt.star) {
+    // export def, * as ns from '...';
+    parser.next();
+    const specifier = parser.startNode();
+    parser.next();
+    parseExportNamespaceAs(parser, node, specifier, exports);
   }
+  // export def, * as ns [, { x, y as z }] from '...';
+  parseExportSpecifiersMaybe(parser, node);
+
+  parseExportFrom(parser, node);
+  return parser.finishNode(node, "ExportNamedDeclaration");
 }
 
-function peekNextType(parser) {
-  return parser.withLookAhead(1, () => parser.type);
+function parseExportImportSpecifiers(parser, node, exports) {
+  // export { x, y as z } [from '...'];
+  node.declaration = null;
+  node.specifiers = parser.parseExportSpecifiers(exports);
+
+  if (parser.isContextual("from")) {
+    parseExportFrom(parser, node, exports);
+  } else {
+    parser.semicolon();
+  }
+  return parser.finishNode(node, "ExportNamedDeclaration");
 }
 
-function isExportDefaultSpecifier() {
-  return this.type === tt.name &&
-    this.withLookAhead(1, isCommaOrFrom);
+function parseExportNamespace(parser, node, exports) {
+  const specifier = parser.startNode();
+
+  node.specifiers = [];
+  parser.next();
+
+  if (! parser.isContextual("as")) {
+    // export * from '...';
+    parseExportFrom(parser, node);
+    return parser.finishNode(node, "ExportAllDeclaration");
+  }
+  // export * as ns from '...';
+  parseExportNamespaceAs(parser, node, specifier, exports);
+  // export * as ns[, { x, y as z }] from '...';
+  parseExportSpecifiersMaybe(parser, node);
+
+  parseExportFrom(parser, node);
+  return parser.finishNode(node, "ExportNamedDeclaration");
 }
 
-function isCommaOrFrom(parser) {
-  return parser.type === tt.comma ||
-    (parser.type === tt.name &&
-     parser.value === "from");
+function parseExportNamespaceAs(parser, node, specifier, exports) {
+  parser.expectContextual("as");
+  specifier.exported = parser.parseIdent(true);
+  node.specifiers.push(
+    parser.finishNode(specifier, "ExportNamespaceSpecifier")
+  );
+
+  exports[specifier.exported.name] = true;
 }
 
-function parseExportSpecifiersMaybe(node) {
-  if (this.eat(tt.comma)) {
+function parseExportSpecifiersMaybe(parser, node) {
+  if (parser.eat(tt.comma)) {
     node.specifiers.push.apply(
       node.specifiers,
-      this.parseExportSpecifiers()
+      parser.parseExportSpecifiers()
     );
   }
 }
 
-function parseExportFrom(node, exports) {
-  const hasFrom = this.eatContextual("from") && this.type === tt.string;
-  node.source = hasFrom ? this.parseExprAtom() : null;
-
-  if (node.specifiers) {
-    for (let i = 0; i < node.specifiers.length; i++) {
-      const s = node.specifiers[i];
-      const exported = s.exported;
-      exports[exported.name] = true;
-    }
-  }
-
-  this.semicolon();
+function peekNextType(parser) {
+  return withLookAhead(parser, 1, () => parser.type);
 }
 
-function shouldParseExportDeclaration() {
-  return this.type.keyword === "var" ||
```

Comment:
The `shouldParseExportDeclaration` is more accurate than `shouldParseExportStatement`, though.